### PR TITLE
Feature/add ids to names mapping

### PR DESCRIPTION
**Pull request by mikeisen1**
Pull data for and add mapping from flow ids to flow names and node ids to node names, as that mapping is not provided by the Elasticsearch data.

### CODE REVIEW

File: kibana/idvametrics/analytics_script.py [hunk 1]
```diff
@@ -262,15 +264,79 @@ def send_query_and_evaluate_result(
...
+    return {"Authorization": f"Bearer {access_token}"}
+
+
+def get_mappings(flow_id: str, email: str, password: str, login_url: str, flows_url: str):
+    """
+    Returns the mappings of flow id to flow name and node id to node names.
+    """
+    header = get_authorization_header_to_idva_flows(email, password, login_url)
+    # pulling out the flow data for all flows
+    flows = requests.get(flows_url, headers=header).json()["flowsInfo"]
```
Comment (mikeisen1) [inline kibana/idvametrics/analytics_script.py:285]: @folksgl and @dzaslavskiy, which is the better approach in the case of a `KeyError`: Stop running the script or have flow/node names be `None`?

File: kibana/manifest.yaml [hunk 2]
```diff
@@ -25,3 +25,7 @@ applications:
         --host https://identity-idva-es-proxy-((ENVIRONMENT_NAME)).apps.internal \
         --port 61443
         --flow_id ((FLOW_ID))
+        --email ((EMAIL))
+        --password ((PASSWORD))
+        --login_url https://idva-api-((ENVIRONMENT_NAME)).app.cloud.gov/v1/customers/login
+        --flows_url https://idva-api-((ENVIRONMENT_NAME)).app.cloud.gov/v1/flows
```
Comment (dzaslavskiy) [inline kibana/manifest.yaml:31]: the arg here should be base_url. the paths are fixed

**Review by dzaslavskiy — CHANGES_REQUESTED** [review]

File: kibana/manifest.yaml [hunk 3]
```diff
@@ -25,3 +25,7 @@ applications:
         --host https://identity-idva-es-proxy-((ENVIRONMENT_NAME)).apps.internal \
         --port 61443
         --flow_id ((FLOW_ID))
+        --email ((EMAIL))
```
Comment (dzaslavskiy) [inline kibana/manifest.yaml:28]: this should be username or user id

**Comment by mikeisen1** [comment]
@folksgl, in the manifest, when we create this script as a worker process under the kibana app, we are still able to run the script every once in a while. Per the [documentation](https://cli.cloudfoundry.org/en-US/v8/run-task.html), using `--process` in the `run-task` command will run the script as a one-off task.

File: kibana/manifest.yaml [hunk 4]
```diff
@@ -25,3 +25,7 @@ applications:
         --host https://identity-idva-es-proxy-((ENVIRONMENT_NAME)).apps.internal \
         --port 61443
         --flow_id ((FLOW_ID))
+        --username ((USERNAME))
+        --password ((PASSWORD))
+        --login_url https://idva-api-((ENVIRONMENT_NAME)).app.cloud.gov/v1/customers/login
+        --base_url https://idva-api-((ENVIRONMENT_NAME)).app.cloud.gov/v1/flows
```
Comment (dzaslavskiy) [inline kibana/manifest.yaml:31]: There should only be one arg here: base_url

**Review by dzaslavskiy — CHANGES_REQUESTED** [review]

File: kibana/idvametrics/analytics_script.py [hunk 5]
```diff
@@ -262,15 +263,75 @@ def send_query_and_evaluate_result(
...
 
+def get_authorization_header_to_idva_flows(email: str, password: str, login_url: str):
+    """
+    Returns a header, containing an access token, for an http request to flow data.
+    """
+    # the required authentication to obtain an access token
+    auth = {"email": email, "password": password}
+
+    # the required access token for accessing flow data
+    access_token = requests.post(login_url, data=auth).json()["access_token"]
```
Comment (dzaslavskiy) [inline kibana/idvametrics/analytics_script.py:274]: handle non success response

File: kibana/manifest.yaml [hunk 6]
```diff
@@ -25,3 +25,6 @@ applications:
         --host https://identity-idva-es-proxy-((ENVIRONMENT_NAME)).apps.internal \
         --port 61443
         --flow_id ((FLOW_ID))
+        --username ((USERNAME))
+        --password ((PASSWORD))
+        --base_url https://idva-api-((ENVIRONMENT_NAME)).app.cloud.gov/v1
```
Comment (dzaslavskiy) [inline kibana/manifest.yaml:30]: please move the /v1 into specific url and out of the base url

**Review by dzaslavskiy — CHANGES_REQUESTED** [review]

**Review by dzaslavskiy — APPROVED** [review]

**Review by folksgl — CHANGES_REQUESTED** [review]
> @folksgl, in the manifest, when we create this script as a worker process under the kibana app, we are still able to run the script every once in a while. Per the [documentation](https://cli.cloudfoundry.org/en-US/v8/run-task.html), using `--process` in the `run-task` command will run the script as a one-off task.

@mikeisen1, @dzaslavskiy I still don't understand why this is being added as a worker process at all. Perhaps I'm missing something, in which case please walk me through the logic here.

My understanding of worker processes is that they're supposed to be continually doing some useful work such as processing a queue or similar backend work that has to happen, but doesn't require accepting and processing web requests directly. **However**, this looks like it will run once and then exit. The entry points to the script are essentially the `get_mappings()` and `send_query_and_evaluate_result()` functions, neither of which contain any logic to continually execute after being run the first time as far as I can tell. If this script is not doing doing _**continuous**_ work, then its essentially just a CF Task, and has no reason to be cemented as a worker process.

**Comment by dzaslavskiy** [comment]
@folksgl Perhaps I am misunderstanding some aspect of the CF interface, but I don't think there is another way to set up a process that then runs on demand. It is being invoked as a task and does not run continuously. It is present in the manifest in order set it up for later execution.

**Comment by folksgl** [comment]
@dzaslavskiy Can the following not completely replace the worker process in the manifest? I haven't run this myself, but it seems to be possible to run commands as tasks without having to set up any extra cf resources in the `kibana` app so long as the command can be run on the container as-is.
`cf run-task kibana --command '~/path/to/analytics-script ...'`

Using `run-task` like this would probably require kibana to add the `python_buildpack` as well which I guess is less than ideal. But I'm also not sure what happens to a worker process that just exits -- does it get re-started non-stop? Is it marked as crashed after so many attempts?

If we do want an app that's just for running this task, it should likely be a separate app (whether in the same manifest or not) and follow the [Run a Task on an App with cf CLI v7](https://docs.cloudfoundry.org/devguide/using-tasks.html#run-tasks) steps in the CF docs, using `cf push APP-NAME --task` to set the process up for repeated execution.

**Comment by dzaslavskiy** [comment]
Yes you can just run the command but that seems awkward and inconvenient. The worker process is started at 0 instances (stopped) and stays that way. the task just runs of the worker manifest. separate manifest is how I did it before with sk.

**Review by folksgl — APPROVED** [review]

**Pull request merged**